envoy/http/header_validator.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+#include "envoy/server/factory_context.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Interface for header key formatters that are stateful. A formatter is created during decoding
+ * headers, attached to the header map, and can then be used during encoding for reverse
+ * translations if applicable.
+ */
+class HeaderValidator {
+public:
+  virtual ~HeaderValidator() = default;
+
+  enum class HeaderEntryValidationResult { Accept, Reject };
+
+  /**
+   * Method for validating a header entry. This is called for both request and response headers.
+   * Returning the Reject value causes the request to be rejected with the 400 status.
+   */
+  virtual HeaderEntryValidationResult validateHeaderEntry(const HeaderString& key,
+                                                          const HeaderString& value) PURE;
+
+  /**
+   * Validate the entire request header map.
+   * This method may mutate header map as well, for example by normalizing URI path.
+   * Returning the Reject value form this method causes HTTP requests to be rejected with 400
+   * status, and gRPC requests with the the INTERNAL (13) error code. Returning the Redirect value
```

Comment:
is INTERNAL consistent? 
looking at https://grpc.github.io/grpc/core/md_doc_statuscodes.html I think INVALID_ARGUMENT makes more sense for bad user headers

Reply:
I used this https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md for mapping HTTP response status from intermediaries to gRPC code.

Reply:
huh.  still seems incorrect but I agree you're doing the officially documented thing and it's probably impossible to change gRPC.